### PR TITLE
impl(generator): support apiVersion in discovery documents

### LINKED ISSUE #13981: proto generation from Discovery Doc

### PR DESCRIPTION
**Pull request by scotthart**
While the Discovery Document schema allows for per-method apiVersion specification, Discovery Documents intended for proto generation are expected to have methods that all have the same value (or lack of value) for `apiVersion`. It is considered an error if that is not the case.

The generator checks early on to make sure the resource definition is well-formed, and the later uses the apiVersion (if present) when formatting the proto file. As verifying the apiVersion requires iterating through all the methods, we want to save that value for later to avoid repeating that processing.

fixes #13981

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13982)

### CODE REVIEW

File: generator/internal/discovery_resource_test.cc [hunk 1]
```diff
@@ -58,6 +59,80 @@ TEST_F(DiscoveryResourceTest, RequiresLROImport) {
...
+      "apiVersion": "test-api-version"
+    }
+  }
+})""";
+  auto resource_json = nlohmann::json::parse(kResourceJson, nullptr, false);
+  ASSERT_TRUE(resource_json.is_object());
+  DiscoveryResource r("myTests", "", resource_json);
+  auto service_api_version = r.GetServiceApiVersion();
+  ASSERT_STATUS_OK(service_api_version);
+  EXPECT_THAT(*service_api_version, Eq("test-api-version"));
```
Comment (dbolduc) [inline generator/internal/discovery_resource_test.cc:99]: same w.r.t. `IsOkAndHolds(...)`

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13982)* status: 0 of 7 files reviewed, 8 unresolved discussions (waiting on @dbolduc)

___
*[`generator/internal/discovery_resource.cc` line 94 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13982#gh-1570981141:-Nvmx6ptCqyhD7cRHvEr:b-18gej2) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/71094e49901b0323abbbad9eacf4ffc94a97ba72/generator/internal/discovery_resource.cc#L94)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

do we have to synchronize `service_api_version_`? (I don't know if the surrounding code is async)

I think the answer is no because
- we only set it immediately before returning, so races are unlikely
- in the case of a race, we do some extra work and find the same answer. No big deal.
</blockquote></details>

You bring up a good point that I hadn't considered. I think the best path forward is to skip the lazy initialization of the mutable member variable and just make `Get` and `Set` separate functions.
___
*[`generator/internal/discovery_resource.cc` line 110 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13982#gh-1570976698:-NvmxOgc4ZNRbh9YxWk1:b-7vybwr) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/71094e49901b0323abbbad9eacf4ffc94a97ba72/generator/internal/discovery_resource.cc#L110)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

The list can start with N empty strings, then have a non-empty value and we will use that.
</blockquote></details>

Added test case.
___
*[`generator/internal/discovery_resource.cc` line 118 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13982#gh-1570978755:-NvmyBvTBM2xD_2gKJKo:b-no83lw) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/71094e49901b0323abbbad9eacf4ffc94a97ba72/generator/internal/discovery_resource.cc#L118)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

I am confused as to why we are treating the empty string as a special case. I am not so familiar with `nlohmann::json` but can we do something like:

```cc
auto methods = *json_.find("methods");
auto it = methods.begin();
auto service_api_version = it->value("apiVersion", "");
for (; it != methods.end(); ++it) {
  if (service_api_version != it->value("apiVersion", "")) {
      service_api_version_ = internal::InvalidArgumentError(
          "resource contains methods with different apiVersion values",
          GCP_ERROR_INFO().WithMetadata("json", json_.dump()));
      return *service_api_version_;
  }
}
```
</blockquote></details>

Removed the temporary string variable and just using the member variable directly.
___
*[`generator/internal/discovery_resource_test.cc` line 76 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13982#gh-1570984092:-NvmxVBPASgqv9CvVC40:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/71094e49901b0323abbbad9eacf4ffc94a97ba72/generator/internal/discovery_resource_test.cc#L76)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

same
</blockquote></details>

Done.
___
*[`generator/internal/discovery_resource_test.cc` line 99 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13982#gh-1570984850:-NvmxWjX75hd4pb8vtzn:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/71094e49901b0323abbbad9eacf4ffc94a97ba72/generator/internal/discovery_resource_test.cc#L99)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

same w.r.t. `IsOkAndHolds(...)`
</blockquote></details>

Done.
___
*[`generator/internal/discovery_resource_test.cc` line 756 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13982#gh-1570986923:-Nvmxmjx84khkGupgv8g:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/71094e49901b0323abbbad9eacf4ffc94a97ba72/generator/internal/discovery_resource_test.cc#L756)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

same
</blockquote></details>

Done.
___
*[`generator/internal/discovery_to_proto.cc` line 197 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13982#gh-1570988447:-NvmxnhgCK42vZbkttJG:bi8shqx) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/71094e49901b0323abbbad9eacf4ffc94a97ba72/generator/internal/discovery_to_proto.cc#L197)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

style nit: I think we tend to say `!status_or` instead of `!status_or.ok()`
</blockquote></details>

We can do that with `Status` types, but attempting to do so with `StatusOr` 
 types leads to a compile error (I tried it).
___
*[`generator/internal/discovery_to_proto_test.cc` line 300 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13982#gh-1570971465:-Nvmxyjg6T_Oqkmsmt4B:b-n6bipu) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/71094e49901b0323abbbad9eacf4ffc94a97ba72/generator/internal/discovery_to_proto_test.cc#L300)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

I think the test will fail if we switch the order of these.
</blockquote></details>

Expanded test to cover this case.

**Review by dbolduc — COMMENTED** [review]
> Reviewed 2 of 6 files at r2.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13982)* status: 2 of 7 files reviewed, 10 unresolved discussions (waiting on @scotthart)

___
*[`generator/internal/discovery_resource.h` line 60 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13982#-Nvn656p7wKn8BZhlUrM:-Nvn656p7wKn8BZhlUrN:bcmwurj) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/141c892e14ff6b139ff499dc4e6b00d8e7e92aff/generator/internal/discovery_resource.h#L60)):*
> ```objc
>   // generation.
>   StatusOr<std::string> GetServiceApiVersion() const;
>   Status SetServiceApiVersion();
> ```

comment: It's a little strange for the `Set*` to return a `Status` when that is also what `GetServiceApiVersion().status()` would yield. But it's ok.
___
*[`generator/internal/discovery_resource.cc` line 118 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13982#gh-1570978755:-Nvn3pRk8pptP3hQbhID:b-5q82m9) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/71094e49901b0323abbbad9eacf4ffc94a97ba72/generator/internal/discovery_resource.cc#L118)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

Removed the temporary string variable and just using the member variable directly.
</blockquote></details>

Strongly consider:

```cc
if (!service_api_version_) service_api_version_ = method_api_version;
if (**service_api_version_ != method_api_version) {
        service_api_version_ = internal::InvalidArgumentError(
          "resource contains methods with different apiVersion values",
          GCP_ERROR_INFO().WithMetadata("json", json_.dump()));
      return service_api_version_->status();
}
```
___
*[`generator/internal/discovery_resource.cc` line 274 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13982#-Nvn5ZUzC5PPpnH-sanr:-Nvn5ZUzC5PPpnH-sans:b6bxh1l) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/141c892e14ff6b139ff499dc4e6b00d8e7e92aff/generator/internal/discovery_resource.cc#L274)):*
> ```cpp
> StatusOr<std::string> DiscoveryResource::JsonToProtobufService(
>     DiscoveryDocumentProperties const& document_properties) const {
>   if (!service_api_version_.has_value()) {
> ```

This check belongs in `GetServiceApiVersion(...)`.

nit: I think we prefer just `if (!service_api_version_)` for optionals.
___
*[`generator/internal/discovery_resource.cc` line 290 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13982#-Nvn5kpN18_GIw6hK3EP:-Nvn5kpN18_GIw6hK3EQ:b-fgo4q4) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/141c892e14ff6b139ff499dc4e6b00d8e7e92aff/generator/internal/discovery_resource.cc#L290)):*
> ```cpp
>                       document_properties.default_hostname));
>   auto service_api_version = GetServiceApiVersion();
>   if (service_api_version.ok() && !service_api_version->empty()) {
> ```

same nit about `status_or.ok()`
___
*[`generator/internal/discovery_resource_test.cc` line 75 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13982#-Nvn8tU9CVOPKqtQhFZy:-Nvn8tU9CVOPKqtQhFZz:b-i7m9fw) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/141c892e14ff6b139ff499dc4e6b00d8e7e92aff/generator/internal/discovery_resource_test.cc#L75)):*
> ```cpp
>   ASSERT_TRUE(resource_json.is_object());
>   DiscoveryResource r("myTests", "", resource_json);
>   ASSERT_STATUS_OK(r.SetServiceApiVersion());
> ```

nit: s/ASSERT/EXPECT/ for all of these. We do not use the value.
___
*[`generator/internal/discovery_to_proto.cc` line 197 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13982#gh-1570988447:-Nvn6ho02QjIVt82Tx8u:bq3lro5) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/71094e49901b0323abbbad9eacf4ffc94a97ba72/generator/internal/discovery_to_proto.cc#L197)):*
> We can do that with `Status` types, but attempting to do so with `StatusOr`  
> types leads to a compile error (I tried it).

FTR, that is backwards. You probably tried after changing the return type from `StatusOr<>` to `Status` in this commit.

https://github.com/googleapis/google-cloud-cpp/blob/e25cfa5e6d5be933323ae51ec1502b162222e584/google/cloud/status_or.h#L199-L200

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13982)* status: 1 of 7 files reviewed, 10 unresolved discussions (waiting on @dbolduc)

___
*[`generator/internal/discovery_resource.h` line 60 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13982#-Nvn656p7wKn8BZhlUrM:-NvnLwXyCrFYQPjoRw3P:b-kmfrkg) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/141c892e14ff6b139ff499dc4e6b00d8e7e92aff/generator/internal/discovery_resource.h#L60)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

comment: It's a little strange for the `Set*` to return a `Status` when that is also what `GetServiceApiVersion().status()` would yield. But it's ok.
</blockquote></details>

We want to exit early if there's an issue with the `apiVersion` in the discovery document. Having two methods allows one to be const and the other non-const so that we don't have any concurrency issues when we thread the file writes.
___
*[`generator/internal/discovery_resource.cc` line 118 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13982#gh-1570978755:-NvnMRXs4mjjSZb45aot:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/71094e49901b0323abbbad9eacf4ffc94a97ba72/generator/internal/discovery_resource.cc#L118)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

Strongly consider:

```cc
if (!service_api_version_) service_api_version_ = method_api_version;
if (**service_api_version_ != method_api_version) {
        service_api_version_ = internal::InvalidArgumentError(
          "resource contains methods with different apiVersion values",
          GCP_ERROR_INFO().WithMetadata("json", json_.dump()));
      return service_api_version_->status();
}
```
</blockquote></details>

Done.
___
*[`generator/internal/discovery_resource.cc` line 274 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13982#-Nvn5ZUzC5PPpnH-sanr:-NvnMSXV-wY4JDD1isab:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/141c892e14ff6b139ff499dc4e6b00d8e7e92aff/generator/internal/discovery_resource.cc#L274)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

This check belongs in `GetServiceApiVersion(...)`.

nit: I think we prefer just `if (!service_api_version_)` for optionals.
</blockquote></details>

Done.
___
*[`generator/internal/discovery_resource.cc` line 290 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13982#-Nvn5kpN18_GIw6hK3EP:-NvnMSz93rReOrTxIhsH:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/141c892e14ff6b139ff499dc4e6b00d8e7e92aff/generator/internal/discovery_resource.cc#L290)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

same nit about `status_or.ok()`
</blockquote></details>

Done.
___
*[`generator/internal/discovery_resource_test.cc` line 75 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13982#-Nvn8tU9CVOPKqtQhFZy:-NvnMTkUALrQTe7cX7UA:b-a2ujfl) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/141c892e14ff6b139ff499dc4e6b00d8e7e92aff/generator/internal/discovery_resource_test.cc#L75)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

nit: s/ASSERT/EXPECT/ for all of these. We do not use the value.
</blockquote></details>

Changed.
___
*[`generator/internal/discovery_to_proto.cc` line 197 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13982#gh-1570988447:-NvnMVbWEQwwdt06wq1X:bf0auqz) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/71094e49901b0323abbbad9eacf4ffc94a97ba72/generator/internal/discovery_to_proto.cc#L197)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

> We can do that with `Status` types, but attempting to do so with `StatusOr`  
> types leads to a compile error (I tried it).

FTR, that is backwards. You probably tried after changing the return type from `StatusOr<>` to `Status` in this commit.

https://github.com/googleapis/google-cloud-cpp/blob/e25cfa5e6d5be933323ae51ec1502b162222e584/google/cloud/status_or.h#L199-L200
</blockquote></details>

Probably. It makes me sad that we don't handle checking for "ok" in a uniform manner between `Status` and `StatusOr`.

**Review by dbolduc — APPROVED** [review]

**Comment by dbolduc** [comment]
> We want to exit early if there's an issue with the `apiVersion` in the discovery document. Having two methods allows one to be const and the other non-const so that we don't have any concurrency issues when we thread the file writes.

Just to clarify, I was thinking of this, which has the same const, non-const semantics. I am not claiming it is any better than what you have.

```cc
StatusOr<std::string> GetThing() const;
void SetThing();

// then we would
SetThing();
auto status = GetThing().status();
```

**Pull request merged**